Rewrite KNXD to use Debian package during installation

Installing optional component knxd fails on RPI3B+ with Buster.
This is probably an issue with upstream repos.

<code>
/bin/bash ../../libtool  --tag=CXX   --mode=link g++  -g -O2 -std=c++0x -Wno-subobject-linkage   -o findknxusb findknxusb.o -lusb-1.0 -lev -lfmt 
libtool: link: g++ -g -O2 -std=c++0x -Wno-subobject-linkage -o findknxusb findknxusb.o  -lusb-1.0 -lev -lfmt
g++ -DHAVE_CONFIG_H -I. -I../..  -I../../src/libserver -I../../src/common -I/usr/include/libusb-1.0 -I/root/knxdbuild/fmt  -g -O2 -std=c++0x -Wno-subobject-linkage -MT usb.o -MD -MP -MF .deps/usb.Tpo -c -o usb.o usb.cpp
In file included from /usr/lib/gcc/arm-linux-gnueabihf/8/include-fixed/limits.h:194,
                 from /usr/lib/gcc/arm-linux-gnueabihf/8/include-fixed/syslimits.h:7,
                 from /usr/lib/gcc/arm-linux-gnueabihf/8/include-fixed/limits.h:34,
                 from /usr/include/libusb-1.0/libusb.h:62,
                 from usb.h:24,
                 from usb.cpp:23:
/root/knxdbuild/fmt/fmt/format.h: In member function ‘void fmt::internal::ArgFormatterBase<Impl, Char>::visit_char(int)’:
/root/knxdbuild/fmt/fmt/format.h:1825:20: error: expected unqualified-id before numeric constant
     const unsigned CHAR_WIDTH = 1;
                    ^~~~~~~~~~
[...few more warnings...]
make[3]: *** [Makefile:501: usb.o] Error 1
make[3]: Leaving directory '/root/knxdbuild/knxd/src/usb'
make[2]: *** [Makefile:394: all-recursive] Error 1
make[2]: Leaving directory '/root/knxdbuild/knxd/src'
make[1]: *** [Makefile:442: all-recursive] Error 1
make[1]: Leaving directory '/root/knxdbuild/knxd'
make: *** [Makefile:378: all] Error 2
Created symlink /etc/systemd/system/multi-user.target.wants/knxd.service → /lib/systemd/system/knxd.service.
Created symlink /etc/systemd/system/network-online.target.wants/knxd.service → /lib/systemd/system/knxd.service.
Removed /etc/systemd/system/multi-user.target.wants/hciuart.service.
Please reboot your device!
OK. Please restart your system now...
</code>

Actually this is a weird error caused by a name conflict between libfmt and glibc. Newer glibc includes define CHAR_WIDTH to 8, resulting in the code `const unsigned 8 = 1`.
See also https://github.com/fmtlib/fmt/pull/616
This is fixed in libfmt 5.0.0, though the script we download is using tags/3.0.0.

The question is how to get a hotfix for openhabian in here, either patching the header file or switching to a newer tag.

--- a/functions/packages.bash
+++ b/functions/packages.bash
@@ -396,27 +396,50 @@ find_setup() {
 ##    knxd_setup()
 ##
 knxd_setup() {
-  local introText="This will install and setup kndx (successor to eibd) as your EIB/KNX IP gateway and router to support your KNX bus system.\\n\\nThis routine was provided by 'Michels Tech Blog': https://bit.ly/3dzeoKh"
-  local successText="Setup was successful.\\n\\nPlease edit '/etc/default/knxd' to meet your interface requirements. For further information on knxd options, please type 'knxd --help'\\n\\nFurther details can be found under: https://bit.ly/3dzeoKh\\n\\nIntegration into openHAB 2 is described here: https://github.com/openhab/openhab/wiki/KNX-Binding"
+  local introText="This will install kndx as your EIB/KNX IP gateway and router to support your KNX bus system.\\n\\nNOTE: Typically, you don't need this if you connect via an IP interface or router to your KNX installation. This package is to turn an USB or serial interface into an IP interface.\n\nNOTE: openHABian changed from building and installing latest source to installing the knxd package provided by several distributions."
+  local missingText="Setup could not find knxd package.\n\nopenHABian changed from building and installing latest source to installing the knxd package provided by several distrubutions. In case you have an installation of openHABian on a custom Linux which does not provide knxd package, you could try to installation routine we used before as described at 'Michels Tech Blog': https://bit.ly/3dzeoKh"
+  local errorText="Installation of knxd package failed, see console log for details."
+  local successText="Installation was successful.\\n\\nPlease edit '/etc/default/knxd' to meet your interface requirements. For further information on knxd options, please type 'knxd --help' or see /usr/share/doc/knxd/.\\n\\nIntegration into openHAB 2 is described here: https://github.com/openhab/openhab/wiki/KNX-Binding"
   local temp
 
   temp="$(mktemp "${TMPDIR:-/tmp}"/openhabian.XXXXX)"
 
-  echo -n "$(timestamp) [openHABian] Beginning setup of EIB/KNX IP Gateway and Router with knxd (https://bit.ly/3dzeoKh)... "
+  echo -n "$(timestamp) [openHABian] Beginning setup of knxd package... "
   if [[ -n $INTERACTIVE ]]; then
-    if (whiptail --title "knxd installation?" --yes-button "Continue" --no-button "Cancel" --yesno "$introText" 10 80); then echo "OK"; else echo "CANCELED"; return 0; fi
+    if (whiptail --title "knxd installation?" --yes-button "Continue" --no-button "Cancel" --yesno "$introText" 15 80); then echo "OK"; else echo "CANCELED"; return 0; fi
   else
     echo "OK"
   fi
 
-  echo -n "$(timestamp) [openHABian] Installing knxd... "
-  # TODO: serve file from the repository
-  if ! cond_redirect wget -O "$temp" https://michlstechblog.info/blog/download/electronic/install_knxd_systemd.sh; then echo "FAILED (fetch installer)"; return 1; fi
-  # NOTE: install_knxd_systemd.sh currently does not give proper exit status for errors, so installer claims success...
-  if cond_redirect bash "$temp"; then rm -f "$temp"; echo "OK (reboot required)"; else echo "FAILED (install)"; return 1; fi
+  if ! apt-cache show knxd &>/dev/null; then
+    echo "FAILED (install)";
+    if [[ -n $INTERACTIVE ]]; then
+      whiptail --title "knxd install failed" --msgbox "$missingText" 15 80
+    fi
+    return 1
+  fi
+
+  echo -n "$(timestamp) [openHABian] Installing package knxd... "
+  if cond_redirect apt-get install --yes knxd; then
+    echo "OK";
+  else
+    echo "FAILED (install)"
+    if [[ -n $INTERACTIVE ]]; then
+      whiptail --title "knxd install failed" --msgbox "$errorText" 15 80
+    fi
+    return 1
+  fi
+
+  # optional package which contains command line tools, it is allowed to fail
+  echo -n "$(timestamp) [openHABian] Installing optional package knxd-tools... "
+  if cond_redirect apt-get install --yes knxd-tools; then
+    echo "OK";
+  else
+    echo "FAILED (optional install)"
+  fi 
 
   if [[ -n $INTERACTIVE ]]; then
-    whiptail --title "Operation Successful!" --msgbox "$successText" 15 80
+    whiptail --title "knxd install sucessful" --msgbox "$successText" 15 80
   fi
 }
 

--- a/functions/packages.bats
+++ b/functions/packages.bats
@@ -43,6 +43,20 @@ teardown_file() {
   echo -e "# ${COL_GREEN}$(timestamp) [openHABian] MQTT service is running.${COL_DEF}" >&3
 }
 
+@test "destructive-knxd_install" {
+  echo -e "# ${COL_CYAN}$(timestamp) [openHABian] knxd installation starting...${COL_DEF}" >&3
+  run knxd_setup 3>&-
+  if [ "$status" -ne 0 ]; then echo "$output" >&3; fi
+  [ "$status" -eq 0 ]
+  echo -e "# ${COL_GREEN}$(timestamp) [openHABian] knxd installation successful.${COL_DEF}" >&3
+
+  # knxd service must be configured and is not started automatically, but we can call knxd executable
+  run knxd --version 3>&-
+  if [ "$status" -ne 0 ]; then echo "$output" >&3; fi
+  [ "$status" -eq 0 ]
+  echo -e "# ${COL_GREEN}$(timestamp) [openHABian] knxd executable ok.${COL_DEF}" >&3
+}
+
 @test "destructive-1wire_install" {
   echo -e "# ${COL_CYAN}$(timestamp) [openHABian] 1wire installation starting...${COL_DEF}" >&3
   run 1wire_setup 3>&-

